content/blog/2025-anndata-012.md

```diff
@@ -0,0 +1,74 @@
++++
+title = "Improved i/o in anndata 0.12"
+date = 2025-07-25T00:00:05+01:00
+description = "With zarr v3 support and xarray integration, anndata takes a big step forward when it comes to i/o tasks."
+author = "Ilan Gold"
+draft = false
++++
+
+# 0.12 released
+
+We're happy to announce that `anndata` 0.12 is out now!
+Check out [the changelog](https://anndata.readthedocs.io/en/stable/release-notes/index.html#v0-12-0) for a full list of changes.
+Here, we want to give our users a bit of a deep dive into the new functionality.
+We have lots of great features, like zarr v3 support (package and format), full lazy loading, and new API customisability! Let’s dive in!
+
+## Zarr v3
+
+[Zarr v3 as a file format](https://zarr-specs.readthedocs.io/en/latest/v3/core/index.html) provides improved cloud support, support for sharding to reduce the number of files created in a zarr store, and improved support for extensibility.
+Check out this graphic from the zarr docs.
+By grouping chunks into shards, which are the individual file units, we can cut down the number of files a given zarr store has.
+Compare this to zarr v2 where every chunk was its own file!
+No more file system slowdowns, you can create large zarr stores with no concerns while still retaining the ease of use of zarr.
+
+<img src="https://zarr.dev/zeps/assets/images/sharding.png" style="max-width: 100%;" alt="zarr v3 sharding graphic" />
+
+
+[The zarr v3 package](https://zarr.readthedocs.io/en/stable/index.html), on top of the new format, provides a fully concurrent + parallel backend for faster cloud and local access over v2, which was both single-threaded and synchronous.
+Thus, the new zarr v3 package on its own should provide a nice speedup, but if that’s not enough, various components are also extendable/replaceable.
```

Comment:
well, so far at least our tests are still slow. I guess that’s just https://github.com/dask/dask/pull/11952, so users shouldn’t run into it?

Reply:
In theory, that should help.  I think the tests is a very-small-d data problem, which also appears with sharding.   It's better not to shard when your data is small, for example!

See https://gist.github.com/ilan-gold/c73383def3798df2724405aa64e40c3d